### PR TITLE
🚀 feat(model): add UniNet

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
## 📝 Description

- Add UniNet model
- `uninet.yaml` does not work as Early Stopping from config is broken for all the models.
- Closes https://github.com/open-edge-platform/anomalib/issues/2781

### Combined

| category   | Original Acc | Original AUROC | Anomalib AUROC | Original F1 | Anomalib F1 |
|:-----------|-------------:|---------------:|---------------:|------------:|------------:|
| bottle     |        100   |         100    |          99.76 |       100   |       98.39 |
| cable      |         98.8 |          97.6  |          98.18 |        94.5 |       93.99 |
| capsule    |        100   |         100    |          94.81 |       100   |       95.89 |
| carpet     |         99.6 |          98.6  |          89.61 |        97.8 |       89.77 |
| grid       |         99.2 |          97.8  |         100    |        95   |       99.12 |
| hazelnut   |        100   |         100    |          99.93 |        99.3 |       98.57 |
| leather    |        100   |         100    |          99.39 |        99.5 |       97.33 |
| metal_nut  |        100   |         100    |         100    |       100   |       99.46 |
| pill       |         99.7 |          98.4  |          96.73 |        99.3 |       97.12 |
| screw      |        100   |         100    |          94.14 |       100   |       91.83 |
| tile       |         98.9 |          96.9  |          98.48 |        94.7 |       95.81 |
| toothbrush |        100   |         100    |          95.28 |       100   |       90.91 |
| transistor |         99.2 |          99.2  |          98.54 |        97.5 |       96.1  |
| wood       |         99.9 |         100    |          99.12 |        98.4 |       96.67 |
| zipper     |         99.7 |          99    |          94.01 |        97.9 |       95.83 |

### Original

| object     |   image_auroc |    acc |     f1 |
|:-----------|--------------:|-------:|-------:|
| carpet     |         98.6  |  99.6  |  97.8  |
| bottle     |        100    | 100    | 100    |
| hazelnut   |        100    | 100    |  99.3  |
| leather    |        100    | 100    |  99.5  |
| cable      |         97.6  |  98.8  |  94.5  |
| capsule    |        100    | 100    | 100    |
| grid       |         97.8  |  99.2  |  95    |
| pill       |         98.4  |  99.7  |  99.3  |
| transistor |         99.2  |  99.2  |  97.5  |
| metal_nut  |        100    | 100    | 100    |
| screw      |        100    | 100    | 100    |
| toothbrush |        100    | 100    | 100    |
| zipper     |         99    |  99.7  |  97.9  |
| tile       |         96.9  |  98.9  |  94.7  |
| wood       |        100    |  99.9  |  98.4  |

### Anomalib

|    | data.init_args.category   |   image_AUROC |   image_F1Score |
|---:|:--------------------------|--------------:|----------------:|
|  0 | bottle    |         99.76 |           98.39 |
|  1 | cable     |         98.18 |           93.99 |
|  2 | capsule   |         94.81 |           95.89 |
|  3 | carpet    |         89.61 |           89.77 |
|  4 | grid      |        100    |           99.12 |
|  5 | hazelnut  |         99.93 |           98.57 |
|  6 | leather   |         99.39 |           97.33 |
|  7 | metal_nut |        100    |           99.46 |
|  8 | pill      |         96.73 |           97.12 |
|  9 | screw     |         94.14 |           91.83 |
| 10 | tile      |         98.48 |           95.81 |
| 11 | toothbrush|         95.28 |           90.91 |
| 12 | transistor|         98.54 |           96.1  |
| 13 | wood      |         99.12 |           96.67 |
| 14 | zipper    |         94.01 |           95.83 |

### Anomaly Map Examples

image
image

## ✨ Changes

Select what type of change your PR is:

- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] ♻️ Refactor (non-breaking change which refactors the code base)
- [ ] ⚡️ Performance improvements
- [ ] 🎨 Style changes (code style/formatting)
- [ ] 🧪 Tests (adding/modifying tests)
- [ ] 📚 Documentation update
- [ ] 📦 Build system changes
- [ ] 🏗️ CI/CD configuration
- [ ] 🔧 Chore (general maintenance)
- [ ] 🔒 Security update
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).
- [ ] 🏷️ My PR title follows conventional commit format.

For more information about code review checklists, see the [Code Review Checklist](https://github.com/open-edge-platform/anomalib/blob/main/docs/source/markdown/guides/developer/contributing.md).

### CODE REVIEW

File: src/anomalib/models/image/uninet/lightning_model.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            source_teacher=source_teacher,
+            target_teacher=target_teacher,
+        )
+
+    def training_step(self, batch: Batch, *args, **kwargs) -> STEP_OUTPUT:
+        """Perform a training step of UniNet."""
+        del args, kwargs  # These variables are not used.
+        source_target_features, student_features, predictions = self.model(batch.image)
+        student_features = self._feature_selection(source_target_features, student_features)
+        loss = self._compute_loss(student_features, source_target_features, predictions, batch.gt_label, batch.gt_mask)
```
Comment (samet-akcay) [inline src/anomalib/models/image/uninet/lightning_model.py]: would it be possible if torch model directly returns loss in training step and predictions in eval steps?

File: src/anomalib/models/components/backbone/de_resnet.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
-"""PyTorch model defining the decoder network for Reverse Distillation.
+"""PyTorch model defining the decoder network for Reverse Distillation and UniNet.
```
Comment (samet-akcay) [inline src/anomalib/models/components/backbone/de_resnet.py]: would it be an idea to have a better module name for this? `de_resnet.py` never made sense to me

File: src/anomalib/models/image/uninet/components/__init__.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+from .attention_bottleneck import AttentionBottleneck, BottleneckLayer
+from .dfs import DomainRelatedFeatureSelection
+
+__all__ = [
+    "DomainRelatedFeatureSelection",
+    "AttentionBottleneck",
+    "BottleneckLayer",
+]
```
Comment (samet-akcay) [inline src/anomalib/models/image/uninet/components/__init__.py]: nit-picky comment... if multiline is not intentional, this can fit into a single line?
```suggestion
__all__ = ["AttentionBottleneck", "BottleneckLayer", "DomainRelatedFeatureSelection"]
```

File: src/anomalib/models/image/uninet/lightning_model.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+from anomalib.models.components import AnomalibModule
+from anomalib.models.components.backbone import get_decoder
+from anomalib.post_processing import PostProcessor
+from anomalib.pre_processing import PreProcessor
+from anomalib.visualization import Visualizer
+
+from .components import (
+    AttentionBottleneck,
+    BottleneckLayer,
+)
```
Comment (samet-akcay) [inline src/anomalib/models/image/uninet/lightning_model.py]:
```suggestion
from .components import AttentionBottleneck, BottleneckLayer
```

File: src/anomalib/models/image/uninet/lightning_model.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        evaluator: Evaluator | bool = True,
+        visualizer: Visualizer | bool = True,
+    ) -> None:
+        super().__init__(
+            pre_processor=pre_processor,
+            post_processor=post_processor,
+            evaluator=evaluator,
+            visualizer=visualizer,
+        )
+        self.loss = UniNetLoss(temperature=temperature)  # base class expects self.loss in lightning module
```
Comment (samet-akcay) [inline src/anomalib/models/image/uninet/lightning_model.py:54]: Maybe we could remove `self.loss` in the base class? I see it is just defined in `__init__.py` and not used ?

Comment (samet-akcay) [reply src/anomalib/models/image/uninet/lightning_model.py:54]: If we remove it from the base class, this part could be simplified quite a bit

File: src/anomalib/models/image/uninet/lightning_model.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+    def training_step(self, batch: Batch, *args, **kwargs) -> STEP_OUTPUT:
+        """Perform a training step of UniNet."""
+        del args, kwargs  # These variables are not used.
+
+        loss = self.model(
+            images=batch.image,
+            masks=batch.gt_mask,
+            labels=batch.gt_label,
+        )
```
Comment (samet-akcay) [inline src/anomalib/models/image/uninet/lightning_model.py]: if multi-line is not intentional, this could be fit into a single line?

File: src/anomalib/models/image/uninet/lightning_model.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    @property
+    def trainer_arguments(self) -> dict[str, Any]:
+        """Does not require any trainer arguments."""
+        return {}
+
+    @property
+    def learning_type(self) -> LearningType:
+        """The model uses one-class learning.
+
+        Though technicaly it suppports multi-class and few_shot as well.
```
Comment (samet-akcay) [inline src/anomalib/models/image/uninet/lightning_model.py]: any idea how `multi-class` and `few-show` can work here?

File: src/anomalib/models/image/uninet/loss.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
+"""Loss functions for UniNet."""
```
Comment (samet-akcay) [inline src/anomalib/models/image/uninet/loss.py]: I see that you created `components` directory here. Would it make sense to move `loss.py` to the components as well? Maybe even `anomaly_map`? Just food for thought

File: src/anomalib/models/image/uninet/torch_model.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        student_features = [
+            student_features[0][0],
+            student_features[1][0],
+            student_features[2][0],
+            student_features[0][1],
+            student_features[1][1],
+            student_features[2][1],
+        ]
+        if self.training:
+            student_features = self._feature_selection(source_target_features, student_features)
```
Comment (samet-akcay) [inline src/anomalib/models/image/uninet/torch_model.py:97]: would it be an idea to rename `self.dfs` to `self.domain_related_feature_selection` or something like that? Then we could remove `self._feature_selection` method?

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Love the implementation! Only few minor comments

File: src/anomalib/models/components/backbone/de_resnet.py (hunk 2 above)
Comment (ashwinvaidya17) [reply src/anomalib/models/components/backbone/de_resnet.py]: Resnet Transpose? Like ConvTranspose. I think de resnet was the name given by the authors of reverse distillation.

Comment (samet-akcay) [reply src/anomalib/models/components/backbone/de_resnet.py]: - `resnet_decoder.py` could also work. 
- Would it make sense to rename `ResNet` to `ResNetDecoder` as well?
- Also the docstring of `ResNet` should be updated with the correct imports I guess. The imports still show that the module is in reverse distillation

File: src/anomalib/models/components/backbone/__init__.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+Example:
+    >>> from anomalib.models.components.backbone import get_decoder
+    >>> decoder = get_decoder()
+
+See Also:
+    - :func:`anomalib.models.components.backbone.de_resnet`:
+        Decoder network implementation
+"""
+
+# Copyright (C) 2025 Intel Corporation
```
Comment (rajeshgangireddy) [inline src/anomalib/models/components/backbone/__init__.py:12]: copyright goes to the top? 
Same for other files as well.

File: src/anomalib/models/image/uninet/components/attention_bottleneck.py [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+    eps = bn.eps
+    std = (running_var + eps).sqrt()
+    t = (gamma / std).reshape(-1, 1, 1, 1)
+    return kernel * t, beta - running_mean * gamma / std
+
+
+class AttentionBottleneck(nn.Module):
+    """Attention Bottleneck for UniNet."""
+
+    expansion: int = 4
```
Comment (rajeshgangireddy) [inline src/anomalib/models/image/uninet/components/attention_bottleneck.py]: can you rename `expansion` to be a bit more specific or add a comment.

File: src/anomalib/models/image/uninet/torch_model.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+    def __init__(
+        self,
+        student: nn.Module,
+        bottleneck: nn.Module,
+        teacher_backbone: str,
+        loss: nn.Module,
+    ) -> None:
+        super().__init__()
+        self.teachers = Teachers(teacher_backbone)
```
Comment (rajeshgangireddy) [inline src/anomalib/models/image/uninet/torch_model.py:48]: Is there any advantage of creating the student and bottleneck in lightning model and the teachers in torch model ?

File: src/anomalib/models/image/uninet/torch_model.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+        self.student = student
+        self.bottleneck = bottleneck
+        self.dfs = DomainRelatedFeatureSelection()
+
+        self.loss = loss
+        self.bce_loss = torch.nn.BCEWithLogitsLoss()
+        # Used to post-process the student features from the de_resnet model to get the predictions
+        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
+        self.fc = nn.Linear(256, 1)
```
Comment (rajeshgangireddy) [inline src/anomalib/models/image/uninet/torch_model.py:57]: would this 256 in nn.Linear(256, 1) be valid for the set of possible backbones?

**Review by rajeshgangireddy — COMMENTED** [review]
Nice work. I have a few small  comments

File: src/anomalib/models/image/uninet/torch_model.py (hunk 13 above)
Comment (ashwinvaidya17) [reply src/anomalib/models/image/uninet/torch_model.py:57]: I just tested by changing the backbone to resnet50. Seems to be working

**Review by rajeshgangireddy — APPROVED** [review]
